src/SourceFile.js

```diff
@@ -93,57 +80,126 @@ class SourceFile extends DirItem {
-                result = rule.match({
-                    ir,
-                    locale: detectedLocale || this.project.getSourceLocale(),
-                    file: this.filePath
-                });
-                if (result) issues = issues.concat(makeArray(result));
-            });
-        });
+        if (!this.filePath) return [];
+        return this.getParsers().flatMap((parser) => {
+            let didWrite = false;
+
+            const /** @type {Result[]} */ fixedResults = [];
+            let /** @type {Result[]} */ currentParseResults = [];
+
+            do {
+                const irs = parser.parse();
+                // indicate that for current intermediate representations, parser did not write out modified representations yet
+                didWrite = false;
+                // clear the results of the current parse, because the file had been overwritten
+                // (so the results don't match the file anymore)
+                currentParseResults = [];
+
+                for (const ir of irs) {
+                    // find the rules that are appropriate for this intermediate representation and then apply them
+                    const rules = this.filetype.getRules().filter((rule) => rule.getRuleType() === ir.getType());
+
+                    // apply rules
+                    const results = rules.flatMap(
+                        (rule) =>
+                            rule.match({
+                                ir,
+                                locale: detectedLocale || this.project.getSourceLocale(),
+                                file: this.filePath,
+                            }) ?? []
+                    );
+
+                    const fixable = results.filter((result) => result.fix !== undefined);
+
+                    let fixer;
+                    if (
+                        // ensure that autofixing is enabled
+                        true === this.project.getConfig().autofix &&
+                        // and that any fixable results were produced
+                        fixable.length > 0 &&
+                        // and that the current parser is able to write
+                        parser.canWrite &&
+                        // and that the fixer for this type of IR is avaliable
+                        (fixer = this.project.getFixerManager().get(ir.getType()))
+                    ) {
+                        // attempt to apply fixes to the current IR
+                        const fixes = /** @type {Fix[]} */ (fixable.map((result) => result.fix));
+                        fixer.applyFixes(ir, fixes);
 
-        return issues;
+                        // check if anything had been applied
+                        if (fixes.some((fix) => fix.applied)) {
+                            // fixer should modify the provided IR
+                            // so tell current parser to write out the modified representation
+                            parser.write(ir);
```

Comment:
maybe write() should return a boolean on whether it was successfully written out? and then didWrite could just be the result of calling parser.write(ir) ?

Reply:
if write() fails for any reason, it should throw - it does not make sense for write() to finish but not write.

didWrite on the other hand is just a flag that indicates that a modification had occurred, so that the while loop would know to reexecute. If You prefer, we could rename it to didApplyFixes or something like that. We could also change the for-of loop over IRs to another do while, so that both of them would clearly depend on whether any fixes have been applied, if You think this would make more sense